### ISSUE TITLE
Crash after undo some actions.

### ISSUE THREAD

**Issue by Kildor** [Beep6581/RawTherapee#3327]
I have a repeteable crash in RT 4.2.1005

I open a photo (Default profile).
Next I change some sliders and curve in Exposition tab.
Switch to the Noise reduction tool (Switched off) and change Chrominance curve.

And now, if I press CtrlZ, the RT crashes with this error:
(rawtherapee.exe:5844): glibmm-ERROR **: 
unhandled exception (type std::exception) in signal handler:
what: vector::_M_range_check: __n (which is 2) >= this->size() (which is 2)

If crash is not happen from the first time, you can try again changing the curve and undo.

Branch: gtk3
Version: 4.2.1005
Changeset: 20fc105cf8494b58fcf55538f00d12707f5cd70d
Compiler: gcc 5.3.0
Processor: generic x86
System: Windows
Bit depth: 64 bits
Gtkmm: V3.18.0
Build type: release
Build flags: -m64 -mwin32 -mthreads -std=gnu++11 -mtune=generic -Werror=unused-label -fopenmp -Werror=unknown-pragmas -msse -msse2 -mwindows -fopenmp -Wno-aggressive-loop-optimizations -DNDEBUG -O3
Link flags: -m64 -mwin32 -mthreads -static-libgcc -mtune=generic -mwindows -s -O3
OpenMP support: ON
MMAP support: ON

**Comment by Floessie** [comment]
@Kildor Good catch, not easy to reproduce.

@Beep6581 Also affects master on Linux.

Backtrace:
```
#0  0x00007ffff1fc613d in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x00007ffff1fedda5 in std::__throw_out_of_range_fmt(char const*, ...) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#2  0x0000000000899dd2 in std::vector<double, std::allocator<double> >::_M_range_check (this=0x2313808, __n=3) at /usr/include/c++/5/bits/stl_vector.h:803
#3  0x0000000000899525 in std::vector<double, std::allocator<double> >::at (this=0x2313808, __n=3) at /usr/include/c++/5/bits/stl_vector.h:824
#4  0x0000000000a07173 in MyFlatCurve::draw (this=0x23135b0) at /home/user/src/rawtherapee/rtgui/myflatcurve.cc:313
#5  0x0000000000a08fc8 in MyFlatCurve::handleEvents (this=0x23135b0, event=0x7fffffffcf10) at /home/user/src/rawtherapee/rtgui/myflatcurve.cc:604
#6  0x0000000000a114b0 in sigc::bound_mem_functor1<bool, MyFlatCurve, _GdkEvent*>::operator() (this=0x23139f8, _A_a1=@0x7fffffffc9b8: 0x7fffffffcf10)
    at /usr/include/sigc++-2.0/sigc++/functors/mem_fun.h:2064
#7  0x0000000000a112eb in sigc::adaptor_functor<sigc::bound_mem_functor1<bool, MyFlatCurve, _GdkEvent*> >::operator()<_GdkEvent* const&> (this=0x23139f0, 
    _A_arg1=@0x7fffffffc9b8: 0x7fffffffcf10) at /usr/include/sigc++-2.0/sigc++/adaptors/adaptor_trait.h:89
#8  0x0000000000a1101e in sigc::internal::slot_call1<sigc::bound_mem_functor1<bool, MyFlatCurve, _GdkEvent*>, bool, _GdkEvent*>::call_it (rep=0x23139c0, 
    a_1=@0x7fffffffc9b8: 0x7fffffffcf10) at /usr/include/sigc++-2.0/sigc++/functors/slot.h:149
#9  0x00007ffff465b68b in ?? () from /usr/lib/x86_64-linux-gnu/libgtkmm-2.4.so.1
#10 0x00007ffff63bb99c in ?? () from /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0
#11 0x00007ffff727dfa5 in g_closure_invoke () from /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#12 0x00007ffff7290264 in ?? () from /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#13 0x00007ffff72987f9 in g_signal_emit_valist () from /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#14 0x00007ffff729908f in g_signal_emit () from /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#15 0x00007ffff64d323f in ?? () from /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0
#16 0x00007ffff63ba752 in gtk_main_do_event () from /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0
#17 0x00007ffff6015baf in ?? () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#18 0x00007ffff6015b55 in ?? () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#19 0x00007ffff6015b55 in ?? () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#20 0x00007ffff6015b55 in ?? () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#21 0x00007ffff6015b55 in ?? () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#22 0x00007ffff6015b55 in ?? () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#23 0x00007ffff6012681 in ?? () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#24 0x00007ffff6012fb8 in gdk_window_process_all_updates () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#25 0x00007ffff633f9b1 in ?? () from /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0
#26 0x00007ffff5ff1d67 in ?? () from /usr/lib/x86_64-linux-gnu/libgdk-x11-2.0.so.0
#27 0x00007ffff6fa705a in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#28 0x00007ffff6fa7400 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#29 0x00007ffff6fa7722 in g_main_loop_run () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#30 0x00007ffff63b9567 in gtk_main () from /usr/lib/x86_64-linux-gnu/libgtk-x11-2.0.so.0
#31 0x00007ffff45f192f in Gtk::Main::run(Gtk::Window&) () from /usr/lib/x86_64-linux-gnu/libgtkmm-2.4.so.1
#32 0x00000000009d1302 in main (argc=1, argv=0x7fffffffe2c8) at /home/user/src/rawtherapee/rtgui/main.cc:361
```

Fix:
```patch
index fe53565..0bc3bd5 100644
--- a/rtgui/myflatcurve.cc
+++ b/rtgui/myflatcurve.cc
@@ -1807,6 +1807,7 @@ void MyFlatCurve::setPoints (const std::vector<double>& p)
     stopNumericalAdjustment();
     FlatCurveType t = (FlatCurveType)p[ix++];
     curve.type = t;
+    lit_point = -1;
 
     if (t == FCT_MinMaxCPoints) {
         curve.x.clear ();
```

@heckflosse @Beep6581 Could one of you countercheck?

**Comment by Kildor** [comment]
I just clicked on wrong widget, unintentionaly change it and tried to undo the changing.
Thanks for fixing $)

**Comment by Beep6581** [comment]
I cannot reproduce the crash in master nor in gtk3.

> Switch to the Noise reduction tool (Switched off) and change Chrominance curve.

You didn't mention turning NR on at any point. I tried reproducing by following the steps verbatim, and also by trying to turn it on before ctrl+z. Neither caused a crash. Tried changing the curve and undoing many times.

Tried the patch despite being unable to reproduce, it didn't cause any issues.

**Comment by Floessie** [comment]
@Beep6581 I found an easy reproduction:

1. Enable NR (the steps on Exposure aren't neccessary)
2. Add points to the chrominance curve
3. Keep mouse over last point (so that you see the vertical line)
4. Ctrl-Z

HTH,
Flössie

**Comment by Kildor** [comment]
> You didn't mention turning NR on at any point.
I didn't turn it on. That was the problem of reproducing.
Yes, keeping mouse on the control point and undo changing this point leads to crash. If you do not switch the tool, you can't undo it, so no crash. But if you've done some other actions, add some points and press CtrlZ, the RT tries to undo this change and crashes.

**Comment by Beep6581** [comment]
@Floessie reproduced with your steps, and fix confirmed.

**Comment by Floessie** [comment]
@Beep6581 Det låtar bra. I saw you added the fix to 21b9b15.

**Comment by Floessie** [comment]
Fixed with in master with 8b5cf38.
